core/commands/refs.go

```diff
@@ -86,6 +89,12 @@ Note: List all referencess recursively by using the flag '-r'.
 			return
 		}
 
+		cached, _, err := req.Option("cached").Bool()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
```

Comment:
here, i would do something like:
```
dserv := n.DAG
if cached {
    dserv = makeOfflineDatastore(n.Blockstore)
}
```

and then for makeOfflineDatastore, do something like we do in merkledag/test/util.go to create a dagservice with only a blockstore and an offline exchange.

Then, pass that `dserv` to the RefWriter instead of `n.DAG`

That way we avoid having split conditionals further down the codepath.